### PR TITLE
Add namespaced ServiceBroker API

### PR DESCRIPTION
**Pull request by eriknelson**
Depends on #1859 (namespaced serviceclasses), #1880 (update broker registry code)

Adds the new namespaced `ServiceBroker` API.

### CODE REVIEW

File: contrib/examples/apiserver/servicebroker.yaml [hunk 1]
```diff
@@ -0,0 +1,16 @@
+apiVersion: servicecatalog.k8s.io/v1beta1
+kind: ServiceBroker
+metadata:
+  name: test-servicebroker
+  namespace: test-ns
+spec:
+  url: http://beefco.de
```
Comment (n3wscott) [inline contrib/examples/apiserver/servicebroker.yaml]: 🤣

File: pkg/apis/servicecatalog/types.go [hunk 2]
```diff
@@ -46,6 +46,30 @@ type ClusterServiceBrokerList struct {
...
+// latest minor version of the v2 major version.
+type ServiceBroker struct {
+	metav1.TypeMeta
+	metav1.ObjectMeta
+
+	Spec   ServiceBrokerSpec
+	Status ServiceBrokerStatus
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
```
Comment (n3wscott) [inline pkg/apis/servicecatalog/types.go]: I think the deepcopy-gen hook only needs to be added once for the top of the file, is that true? If so then I think you can remove these as it is already on line 26. I don't know if it only applies to the following struct or all structs in the file so please educate me.

File: pkg/apis/servicecatalog/types.go [hunk 3]
```diff
@@ -88,6 +112,15 @@ type ClusterServiceBrokerSpec struct {
 	AuthInfo *ClusterServiceBrokerAuthInfo
 }
 
+// ServiceBrokerSpec represents a description of a Broker.
+type ServiceBrokerSpec struct {
```
Comment (n3wscott) [inline pkg/apis/servicecatalog/types.go]: would it be possible to move `ServiceBrokerAuthInfo` into `CommonServiceBrokerSpec` ? or do they have different futures?

File: pkg/apis/servicecatalog/v1beta1/types.go [hunk 4]
```diff
@@ -148,6 +190,42 @@ type ClusterBearerTokenAuthConfig struct {
 	SecretRef *ObjectReference `json:"secretRef,omitempty"`
 }
 
+// ServiceBrokerAuthInfo is a union type that contains information on
```
Comment (n3wscott) [inline pkg/apis/servicecatalog/v1beta1/types.go]: >a union type

I think this comment is wrong, I think a union type looks like:

```
type ServiceBrokerAuthInfo struct {
	BasicAuthConfig
	BearerTokenAuthConfig
}
```

No?

Comment (n3wscott) [reply pkg/apis/servicecatalog/v1beta1/types.go]: What is here is more of a wrapper of the two types

File: pkg/apis/servicecatalog/types.go (hunk 2 above)
Comment (eriknelson) [reply pkg/apis/servicecatalog/types.go]: I might have copied the comment block from the top, can someone confirm? I'd like to know as well. Also, are these hooks documented somewhere?

**Comment by eriknelson** [comment]
@pmorie rebased

File: pkg/apis/servicecatalog/types.go (hunk 3 above)
Comment (pmorie) [reply pkg/apis/servicecatalog/types.go]: They’re different and have different futures

File: pkg/apis/servicecatalog/types.go (hunk 2 above)
Comment (carolynvs) [reply pkg/apis/servicecatalog/types.go]: I believe that you want deepcopy gen for objects for anything that could be watched for individually, i.e. a root level object. So yes for ServiceBroker, but not necessary for say ServiceBrokerStatus. It helps people make a copy before mutating an object pulled from the cache.

**Pull request merged**